presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageUtils.java

```diff
@@ -0,0 +1,977 @@
+            return buildBlockFromDateMilliVector((DateMilliVector) vector, type);
+        }
+        else if (vector instanceof TimeMilliVector) {
+            return buildBlockFromTimeMilliVector((TimeMilliVector) vector, type);
+        }
+        else if (vector instanceof TimeSecVector) {
+            return buildBlockFromTimeSecVector((TimeSecVector) vector, type);
+        }
+        else if (vector instanceof TimeStampSecVector) {
+            return buildBlockFromTimeStampSecVector((TimeStampSecVector) vector, type);
+        }
+        else if (vector instanceof TimeMicroVector) {
+            return buildBlockFromTimeMicroVector((TimeMicroVector) vector, type);
+        }
+        else if (vector instanceof TimeStampMilliTZVector) {
+            return buildBlockFromTimeMilliTZVector((TimeStampMilliTZVector) vector, type);
+        }
+        else if (vector instanceof ListVector) {
+            return buildBlockFromListVector((ListVector) vector, type);
+        }
+
+        throw new UnsupportedOperationException("Unsupported vector type: " + vector.getClass().getSimpleName());
+    }
+
+    public static Block buildBlockFromDictionaryVector(FieldVector fieldVector, FieldVector dictionaryVector)
+    {
+        // Validate inputs
+        requireNonNull(fieldVector, "encoded vector is null");
+        requireNonNull(dictionaryVector, "dictionary vector is null");
+
+        // Create a BlockBuilder for the decoded vector's data type
+        Type prestoType = getPrestoTypeFromArrowType(dictionaryVector.getField().getType());
+
+        Block dictionaryblock = null;
+        // Populate the block dynamically based on vector type
+        for (int i = 0; i < dictionaryVector.getValueCount(); i++) {
+            if (!dictionaryVector.isNull(i)) {
+                dictionaryblock = appendValueToBlock(dictionaryVector, prestoType);
+            }
+        }
+
+        return getDictionaryBlock(fieldVector, dictionaryblock);
+
+        // Create the Presto DictionaryBlock
+    }
+
+    private static DictionaryBlock getDictionaryBlock(FieldVector fieldVector, Block dictionaryblock)
+    {
+        if (fieldVector instanceof IntVector) {
+            // Get the Arrow indices vector
+            IntVector indicesVector = (IntVector) fieldVector;
+            int[] ids = new int[indicesVector.getValueCount()];
+            for (int i = 0; i < indicesVector.getValueCount(); i++) {
+                ids[i] = indicesVector.get(i);
+            }
+            return new DictionaryBlock(ids.length, dictionaryblock, ids);
+        }
+        else if (fieldVector instanceof BigIntVector) {
+            // Get the BigInt indices vector
+            BigIntVector bigIntIndicesVector = (BigIntVector) fieldVector;
+            int[] ids = new int[bigIntIndicesVector.getValueCount()];
+            for (int i = 0; i < bigIntIndicesVector.getValueCount(); i++) {
+                ids[i] = (int) bigIntIndicesVector.get(i);
```

Comment:
down casting big int to int will result in data loss, right? so better we don't support creating dictionaryblock from bigint indices

Reply:
yeah okay will remove that

Reply:
done